src/vs/workbench/contrib/search/browser/notebookSearch/notebookSearchService.ts

```diff
@@ -0,0 +1,212 @@
+
+		let numResults = 0;
+		for (const result of results) {
+			if (textQuery.maxResults && numResults >= textQuery.maxResults) {
+				limitHit = true;
+				break;
+			}
+
+			if (!scannedFiles.has(result.resource) && !uniqueResults.has(result.resource)) {
+				uniqueResults.set(result.resource, result.cellResults.length > 0 ? result : null);
+				numResults++;
+			}
+		}
+
+		const end = Date.now();
+		this.logService.trace(`query: ${textQuery.contentPattern.pattern}`);
+		this.logService.trace(`closed notebook search time | ${end - start}ms`);
+
+		return {
+			results: uniqueResults,
+			limitHit
+		};
+	}
+
+	private async getLocalNotebookResults(query: ITextQuery, token: CancellationToken, widgets: Array<NotebookEditorWidget>, searchID: string): Promise<IOpenNotebookSearchResults> {
+		const localResults = new ResourceMap<INotebookFileMatchWithModel | null>(uri => this.uriIdentityService.extUri.getComparisonKey(uri));
+		let limitHit = false;
+
+		for (const widget of widgets) {
+			if (!widget.viewModel) {
+				continue;
+			}
+			const askMax = isNumber(query.maxResults) ? query.maxResults + 1 : Number.MAX_SAFE_INTEGER;
+			let matches = await widget
+				.find(query.contentPattern.pattern, {
+					regex: query.contentPattern.isRegExp,
+					wholeWord: query.contentPattern.isWordMatch,
+					caseSensitive: query.contentPattern.isCaseSensitive,
+					includeMarkupInput: query.contentPattern.notebookInfo?.isInNotebookMarkdownInput ?? true,
+					includeMarkupPreview: query.contentPattern.notebookInfo?.isInNotebookMarkdownPreview ?? true,
+					includeCodeInput: query.contentPattern.notebookInfo?.isInNotebookCellInput ?? true,
+					includeOutput: query.contentPattern.notebookInfo?.isInNotebookCellOutput ?? true,
+				}, token, false, true, searchID);
+
+
+			if (matches.length) {
+				if (askMax && matches.length >= askMax) {
+					limitHit = true;
+					matches = matches.slice(0, askMax - 1);
+				}
+				const cellResults: INotebookCellMatchWithModel[] = matches.map(match => {
+					const contentResults = contentMatchesToTextSearchMatches(match.contentMatches, match.cell);
+					const webviewResults = webviewMatchesToTextSearchMatches(match.webviewMatches);
+					return {
+						cell: match.cell,
+						index: match.index,
+						contentResults: contentResults,
+						webviewResults: webviewResults,
+					};
+				});
+
+				const fileMatch: INotebookFileMatchWithModel = {
+					resource: widget.viewModel.uri, cellResults: cellResults
+				};
+				localResults.set(widget.viewModel.uri, fileMatch);
+			} else {
+				localResults.set(widget.viewModel.uri, null);
+			}
+		}
+
+		return {
+			results: localResults,
+			limitHit
+		};
+	}
+
+
+	private getLocalNotebookWidgets(): Array<NotebookEditorWidget> {
+		const notebookWidgets = this.notebookEditorService.retrieveAllExistingWidgets();
+		return notebookWidgets
+			.map(widget => widget.value)
+			.filter((val): val is NotebookEditorWidget => !!val && !!(val.viewModel));
```

Comment:
why do we check `viewModel` here? Is it because we want to ensure the webview is rendered/initialized?

Reply:
I can do `hasModel` if that's better perf